core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -2019,4 +2021,35 @@ public void testRecoverFromForeignTranslog() throws IOException {
             assertThat(topDocs.totalHits, equalTo(numDocs));
         }
     }
+
+    public void testShardNotAvailableExceptionWhenEngineClosedConcurrently() throws IOException, InterruptedException {
```

Comment:
can we add merge to the test name? Alternatively, maybe we should extend this test to check other (like refresh) operations and make sure what ever they throw is caught by TransportActions.isShardNotAvailable ?